**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант №9

Набор данных: auto.csv

Атрибуты:
...

### 1.Чтение файла (набора данных)

In [1]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd
import numpy as np

df = pd.read_csv('auto.csv', sep=',')

### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [ ]:
# Чтение первых 20 строк с помощью метода head
display(df.head(20))

Name  year  SellingPrice  kmdriven     fuel  \
0                     Maruti 800 AC  2007       60000.0   70000.0   Petrol   
1          Maruti Wagon R LXI Minor  2007      135000.0   50000.0   Petrol   
2              Hyundai Verna 1.6 SX  2012      600000.0  100000.0   Diesel   
3            Datsun RediGO T Option  2017      250000.0   46000.0   Petrol   
4             Honda Amaze VX i-DTEC  2014      450000.0  141000.0   Diesel   
5              Maruti Alto LX BSIII  2007      140000.0  125000.0   Petrol   
6         Hyundai Xcent 1.2 Kappa S  2016      550000.0   25000.0   Petrol   
7         Tata Indigo Grand Petroll  2014      240000.0   60000.0   Petrol   
8          Hyundai Creta 1.6 VTVT S  2015      850000.0   25000.0  Petroll   
9          Maruti Celerio Green VXI  2017      365000.0   78000.0      CNG   
10          Chevrolet Sail 1.2 Base  2015      260000.0   35000.0   Petrol   
11         Tata Indigo Grand Petrol  2014      250000.0  100000.0   Petrol   
12  Toyota Corolla Altis 1.8 VL CVT  2018     1650000.0   25000.0   Petrol   
13                    Maruti 800 AC  2007       60000.0   70000.0   Petrol   
14         Maruti Wagon R LXI Minor  2007      135000.0   50000.0   Petrol   
15             Hyundai Verna 1.6 SX  2012      600000.0  100000.0   Diesel   
16           Datsun RediGO T Option  2017      250000.0   46000.0   Petrol   
17            Honda Amaze VX i-DTEC  2014      450000.0  141000.0   Diesel   
18             Maruti Alto LX BSIII  2007      140000.0  125000.0   Petrol   
19        Hyundai Xcent 1.2 Kappa S  2016      550000.0   25000.0   Petrol   

   seller_Type transmission         owner  
0   Individual       Manual   First Owner  
1   Individual       Manual   First Owner  
2   Individual       Manual   First Owner  
3   Individual       Manual   First Owner  
4   Individual       Manual  Second Owner  
5   Individual       Manual   First Owner  
6   Individual       Manual   First Owner  
7   Individual       Manual  Second Owner  
8   Individual       Manual   First Owner  
9   Individual       Manual   First Owner  
10  Individual       Manual   First Owner  
11  Individual       Manual   First Owner  
12      Dealer    Automatic   First Owner  
13  Individual       Manual   First Owner  
14  Individual       Manual   First Owner  
15  Individual       Manual   First Owner  
16  Individual       Manual   First Owner  
17  Individual       Manual  Second Owner  
18  Individual       Manual   First Owner  
19  Individual       Manual   First Owner

2.2 Оценка данных с помощью метода info.

In [4]:
# Оценка данных с помощью метода info
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4344 entries, 0 to 4343
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Name          4344 non-null   object 
 1   year          4344 non-null   int64  
 2   SellingPrice  4342 non-null   float64
 3   kmdriven      4343 non-null   float64
 4   fuel          4343 non-null   object 
 5   seller_Type   4344 non-null   object 
 6   transmission  4344 non-null   object 
 7   owner         4344 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 271.6+ KB


None

2.3 Оценка данных с помощью метода describe.

In [6]:
# оцените числовые столбцы с помощью describe
pd.set_option('display.float_format', '{:.5f}'.format)
display(df.describe())

year  SellingPrice     kmdriven
count 4344.00000    4342.00000   4343.00000
mean  2013.09231  503998.74090  66193.06332
std      4.21454  578446.45972  46636.64876
min   1992.00000   20000.00000      1.00000
25%   2011.00000  210000.00000  35000.00000
50%   2014.00000  350000.00000  60000.00000
75%   2016.00000  600000.00000  90000.00000
max   2020.00000 8900000.00000 806599.00000


---

**Сделать выводы. Вы должны понимать, что означает тот или иной столбец, чтобы
 ответить на вопросы на защите.**

 ---
 ## Промежуточный вывод о данных
Данные содержат 4344 записей о поддержанных авто с 8 признаками, включая имя владельца, год выпуска машины, цену продажи, пробег, тип топлива, тип продавца, тип коробки передач и тип владельца.

Типы данных: int64: year. float64: SellingPrice, kmdriven.

Категориальные: object: Name, fuel, seller_Type, transmission, owner.

Пропуски: SellingPrice - 2 пропущенных значения, kmdriven и fuel — по 1 пропущенному значению.

Средняя цена продажи равна примерно 504000, а медиана 350000, это говорит о том, что часто встречаются аномально дорогие машины.
Средний пробег машины равен примерно 66200, а медиана 60000, эти цифры приблизительно равны, значит большого количества аномалий не наблюдается.

 2.4 Оценка названий столбцов

In [8]:
# Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.
df.columns

Index(['Name', 'year', 'SellingPrice', 'kmdriven', 'fuel', 'seller_Type',
       'transmission', 'owner'],
      dtype='object')

## Анализ и коррекция названий столбцов
Был проведен анализ названий столбцов на предмет наличия проблем:

- Проверка на пробелы в начале/конце названий
- Проверка на наличие специальных символов и пробелов

После анализа выполнено переименование столбцов для улучшения читаемости и удобства работы:

- Name -> car_name
- year -> release_year
- SellingPrice -> selling_price
- kmdriven -> km_driven
- fuel -> fuel_type
- seller_Type -> seller_type
- transmission -> transmission
- owner -> owner_type

В конце выводится обновленный список названий столбцов для подтверждения изменений.

In [9]:
# переименование столбцов
df.rename(columns={
    'Name': 'car_name',
    'year': 'release_year',
    'SellingPrice': 'selling_price',
    'kmdriven': 'km_driven',
    'fuel': 'fuel_type',
    'seller_Type': 'seller_type',
    'transmission': 'transmission',
    'owner': 'owner_type'
}, inplace=True)

print("Обновленные названия столбцов:")
df.columns

Обновленные названия столбцов:


Index(['car_name', 'release_year', 'selling_price', 'km_driven', 'fuel_type',
       'seller_type', 'transmission', 'owner_type'],
      dtype='object')

### 3. Проверка пропусков

In [10]:
# Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением).

# Проверка пропусков
missing_data = df.isna().sum()
print(missing_data[missing_data > 0])

selling_price    2
km_driven        1
fuel_type        1
dtype: int64


In [15]:
# Заполнение числовых столбцов медианой
median_price = df['selling_price'].median()
median_km_driven = df['km_driven'].median()

df['selling_price'] = df['selling_price'].fillna(median_price)
df['km_driven'] = df['km_driven'].fillna(median_km_driven)
print(f"Пропуски были заполнены медианой, для selling_price {median_price}, для km_driven {median_km_driven}")
    
# Заполнение пропусков модой
mode_fuel_type = df['fuel_type'].mode()[0]
df['fuel_type'] = df['fuel_type'].fillna(mode_fuel_type)
print(f"Пропуски fuel_type были заполнены модой {mode_fuel_type}")

Пропуски были заполнены медианой, для selling_price 350000.0, для km_driven 60000.0
Пропуски fuel_type были заполнены модой Diesel


In [16]:
missing_data = df.isna().sum()
print("Пропуски в данных:")
print(missing_data[missing_data > 0])

Пропуски в данных:
Series([], dtype: int64)


### Выбранные стратегии обработки пропусков:

1. Заполнение числовых данных медианой:
- Этот способ был выбран, так как Медиана более устойчива к выбросам, чем среднее значение и сохраняет общее распределение данных без смещения в сторону экстремальных значений

2. Заполнение категориальных данных модой:
- Этот способ был выбран, так как Мода представляет наиболее часто встречающееся значение и сохраняет исходное распределение категориальных переменных
### Альтернативные методы:
1. Удаление строк с пропусками:
- Имеет существенный недостаток - потеря данных и возможное смещение выборки.

2. Создание отдельной категории "Unknown":
- Этот способ применяется только для категориальных данных и позволяет сохранить информацию о том, что там был пропуск, это может быть полезно, если нам необязательны эти данные, в случае с теми данными которые у нас есть такой способ не подходит.

3. Заполнение числовых данных средним значением:
- Недостаток был описан выше, среднее неустойчиво к выбросам и может не сохранить общее распределение данных, при просмотре методом info было видно, как сильно разнятся среднее и медиана, поэтому медиана и была выбрана


---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с пропусками, а также знать другие способы работы с пропусками, чтобы
 ответить на вопросы на защите.**


 ---

### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [17]:
# Проверка на наличие явных дубликатов
duplicates_count = df.duplicated().sum()
print(f"Количество полных дубликатов в данных: {duplicates_count}")

Количество полных дубликатов в данных: 763


In [18]:
# удалите дубликаты, если они есть
df = df.drop_duplicates()
print(f"Удалено {duplicates_count} дубликатов.")

Удалено 763 дубликатов.


#### Проверка неявных дубликатов

In [22]:
# Проверка на наличие неявных дубликатов
categorical_columns = df.select_dtypes(include=['object']).columns.tolist()

# Анализ каждого категориального столбца
for column in categorical_columns:
    print(f"\nАнализ столбца: {column}")
    # Частота каждого значения
    value_counts = df[column].value_counts()
    display(value_counts)
        


Анализ столбца: car_name


car_name
Maruti Swift Dzire VDI                            54
Maruti Alto 800 LXI                               48
Maruti Alto LXi                                   42
Maruti Alto LX                                    30
Maruti Swift VDI BSIV                             28
                                                  ..
Mahindra Verito Vibe 1.5 dCi D6                    1
Ford EcoSport 1.5 Petrol Titanium Plus AT BSIV     1
Tata Harrier XE                                    1
Hyundai Xcent 1.2 CRDi SX                          1
Fiat Linea Emotion (Diesel)                        1
Name: count, Length: 1492, dtype: int64


Анализ столбца: fuel_type


fuel_type
Diesel      1801
Petrol      1719
CNG           37
LPG           22
Petroll        1
Electric       1
Name: count, dtype: int64


Анализ столбца: seller_type


seller_type
Individual          2836
Dealer               711
Trustmark Dealer      33
Dealeer                1
Name: count, dtype: int64


Анализ столбца: transmission


transmission
Manual       3269
Automatic     312
Name: count, dtype: int64


Анализ столбца: owner_type


owner_type
First Owner             2220
Second Owner             980
Third Owner              289
Fourth & Above Owner      75
Test Drive Car            17
Name: count, dtype: int64

In [23]:
# Создаем копию DataFrame для безопасного изменения
df_clean = df.copy()

# Заменяем 'Petroll' на 'Petrol' с использованием replace
df_clean['fuel_type'] = df['fuel_type'].replace('Petroll', 'Petrol')
print("Исправлена опечатка: с 'Petroll' на 'Petrol'")

print("\nОбновленное распределение значений в fuel_type:")
display(df_clean['fuel_type'].value_counts())

# Заменяем 'Dealeer' на 'Dealer' с использованием replace
df_clean['seller_type'] = df['seller_type'].replace('Dealeer', 'Dealer')
print("Исправлена опечатка: с 'Dealeer' на 'Dealer'")

print("\nОбновленное распределение значений в seller_type:")
display(df_clean['seller_type'].value_counts())

# Теперь присваиваем очищенный DataFrame обратно в df
df = df_clean

Исправлена опечатка: с 'Petroll' на 'Petrol'

Обновленное распределение значений в fuel_type:


fuel_type
Diesel      1801
Petrol      1720
CNG           37
LPG           22
Electric       1
Name: count, dtype: int64

Исправлена опечатка: с 'Dealeer' на 'Dealer'

Обновленное распределение значений в seller_type:


seller_type
Individual          2836
Dealer               712
Trustmark Dealer      33
Name: count, dtype: int64

In [ ]:
# удалите дубликаты, если они есть
subset_columns = ['car_name', 'release_year', 'selling_price', 'km_driven', 'fuel_type', 'transmission', 'owner_type']
implicit_duplicates = df.duplicated(subset=subset_columns, keep=False)

# Количество неявных дубликатов
duplicate_count = implicit_duplicates.sum()
print(f"Количество неявных дубликатов: {duplicate_count}")

# Примеры неявных дубликатов
if duplicate_count > 0:
    print("\nПримеры неявных дубликатов:")
    display(df[implicit_duplicates].sort_values(by=subset_columns).head(duplicate_count))
df = df.drop_duplicates(subset=subset_columns, ignore_index=True)
display(df.info())


Количество неявных дубликатов: 2

Примеры неявных дубликатов:


car_name  release_year  selling_price   km_driven  \
8   Hyundai Creta 1.6 VTVT S          2015   850000.00000 25000.00000   
21  Hyundai Creta 1.6 VTVT S          2015   850000.00000 25000.00000   

   fuel_type seller_type transmission   owner_type  
8     Petrol  Individual       Manual  First Owner  
21    Petrol  Individual       Manual  First Owner

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3580 entries, 0 to 3579
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   car_name       3580 non-null   object 
 1   release_year   3580 non-null   int64  
 2   selling_price  3580 non-null   float64
 3   km_driven      3580 non-null   float64
 4   fuel_type      3580 non-null   object 
 5   seller_type    3580 non-null   object 
 6   transmission   3580 non-null   object 
 7   owner_type     3580 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 223.9+ KB


None

#### Поиск дубликатов был выполнен следующим образом:
- Метод df.duplicated(subset=subset_columns, keep=False) находит все строки, где комбинация указанных столбцов совпадает
- Параметр keep=False помечает все дубликаты (а не только вторые и последующие вхождения)
- df = df.drop_duplicates(subset=subset_columns, ignore_index=True), удаляет дубликаты по признакам и перезаписывает индексы после удаления
#### Анализ результатов выполнен так:
- Подсчитывается количество неявных дубликатов
- При наличии дубликатов выводятся примеры для визуальной проверки
- После удаления дубликатов выводится общая информация о датасете



---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с дубликатами, знать, что такое явные и неявные дубликаты и способы работы с ними, чтобы ответить на вопросы на защите.**


 ---

### 5. Провека типов данных

In [26]:
display(df.dtypes)

car_name          object
release_year       int64
selling_price    float64
km_driven        float64
fuel_type         object
seller_type       object
transmission      object
owner_type        object
dtype: object

In [27]:
# Проверьте типы данных, при необходимости измените типы данных, чтобы они соответствовали действительности.
# Конвертация из float64 в int64 для тех данных, что являются целочисленными
copy_df = df.copy()
for col in ['selling_price', 'km_driven']:
    # Проверка точно ли все числа в столбце целые
    if (copy_df[col] % 1 == 0).all():
        copy_df[col] = copy_df[col].astype('int64')

# Список категориальных столбцов
categorical_columns = ['fuel_type', 'seller_type', 'transmission', 'owner_type']
# Конвертация столбцов с ограниченным количеством данных из object в category
for col in categorical_columns:
    copy_df[col] = copy_df[col].astype('category')

df = copy_df
print("Итоговые типы данных:")
display(df.dtypes)


Итоговые типы данных:


car_name           object
release_year        int64
selling_price       int64
km_driven           int64
fuel_type        category
seller_type      category
transmission     category
owner_type       category
dtype: object

Столбцы selling_price и km_driven указаны как float64, но при просмотре данных видно, что цены указаны только в рублях, без копеек а пройденные километры также только в целочисленном виде, эти столбцы были преобразованы в тип int64. Для категориальных признаков fuel_type, seller_type, transmission и owner_type текущий тип object не оптимален: данные в этих столбцах содержат ограниченный набор значений (например, Manual, Automatic), поэтому они были преобразованы в тип category для экономии памяти и ускорения операций группировки.

---

**Обратите внимание, что во всех вариантах необходимо сделать приведение типов. Будьте готовы на защите аргументировать проверку типов (почему выполнены те или иные преобразования).**


 ---

### 6. Группировка данных

#### Задание 1

*`Группировка - тип коробки передач (transmission) по количеству
разных типов топлива (fuel)`*

In [28]:
# выполните группировку согласно варианту
transmission_fuel_counts = df.groupby('transmission')['fuel_type'].value_counts()
display(transmission_fuel_counts)

C:\Users\fedor\AppData\Local\Temp\ipykernel_22284\2463917226.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  transmission_fuel_counts = df.groupby('transmission')['fuel_type'].value_counts()


transmission  fuel_type
Automatic     Diesel        170
              Petrol        141
              Electric        1
              CNG             0
              LPG             0
Manual        Diesel       1631
              Petrol       1578
              CNG            37
              LPG            22
              Electric        0
Name: count, dtype: int64

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

В результате группировки данных видно, что в изначальном датасете сильно преобладают машины с механической коробкой передач. Электромобили крайне редки и представлены только с коробкой-автомат на механике таких машин нет. В свою очередь на автоматической коробке передач нет машин, которые едут на газу (типы LPG и CNG). Большинство машин едут на дизеле и бензине, пропорция машин на таких видах топлива довольно схожа, что на автоматической, что на механической коробках передач, но все же на автоматических заметно небольшое преобладание дизельных автомобилей.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 2

*`Группировка - тип владельца (owner) по количеству разных коробок
передач (transmission). Создать датафрейм. Переименовать столбец с
количеством в “сount”. Отсортировать по убыванию столбца “count”`*

In [31]:
# выполните группировку согласно варианту
# Группировка по 'owner' и подсчет 'transmission'
owner_transmission_counts = df.groupby('owner_type')['transmission'].value_counts().reset_index(name='count')

# Сортировка по убыванию столбца 'count'
owner_transmission_counts_sorted = owner_transmission_counts.sort_values(by='count', ascending=False)
display(owner_transmission_counts_sorted)

C:\Users\fedor\AppData\Local\Temp\ipykernel_22284\2086206196.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  owner_transmission_counts = df.groupby('owner_type')['transmission'].value_counts().reset_index(name='count')


owner_type transmission  count
0           First Owner       Manual   1992
4          Second Owner       Manual    916
8           Third Owner       Manual    271
1           First Owner    Automatic    227
2  Fourth & Above Owner       Manual     73
5          Second Owner    Automatic     64
9           Third Owner    Automatic     18
6        Test Drive Car       Manual     16
3  Fourth & Above Owner    Automatic      2
7        Test Drive Car    Automatic      1

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

Большинство автомобилей в этой выборке имеют механическую коробку передач, особенно среди первых и вторых владельцев. Три верхние позиции по количеству занимают автомобили с механической коробкой. Количество автомобилей значительно уменьшается по мере увеличения числа предыдущих владельцев. Это ожидаемо, так как чем больше владельцев, тем меньше таких автомобилей на рынке. Также отдельно видно, что машин на тест-драйв очень мало и абсолютное большинство из них с автоматической коробкой передач, скорее всего потому что они новые.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 3

*`Сводная таблица (pivot_table) - средняя стоимость для каждого года
year. Отсортировать по убыванию. Округлить до трёх знаков.`*

In [32]:
# выполните сводную таблицу согласно варианту
# Создание сводной таблицы
data_pivot = df.pivot_table(index=['release_year'], values='selling_price', aggfunc='mean')

# Сортировка и округление значений сводной таблицы
data_pivot = data_pivot.sort_values(by='selling_price', ascending=False)
data_pivot['selling_price'] = data_pivot['selling_price'].round(3)
display(data_pivot)

selling_price
release_year               
2019           865364.16700
2018           820420.30500
2020           804844.40000
2017           740057.59400
2016           638342.70700
2014           512003.05900
2015           510051.16000
2013           433046.28500
2012           379352.02700
2011           296028.29100
2010           275936.88600
2009           235467.42000
1996           225000.00000
2008           177668.17100
2007           165282.04300
1998           165111.11100
2006           154182.76300
2004           126697.34200
2005           114821.44300
2001           109562.43800
1995            95000.00000
1997            93000.00000
2003            87090.90900
2002            86111.11100
2000            81500.00000
1999            73888.88900
1992            50000.00000

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной сводной таблицы?`**

В этих данных наблюдается четкая тенденция к росту цен с уменьшением возраста. Чем новее год выпуска, тем выше средняя цена продажи. Все данные можно разбить на 3 периода: новые, средние и старые.

Годы 2018–2020 имеют среднюю цену выше 800000. Годы 2009–2017 находятся в диапазоне от 235000 до 740000. Годы до 2009 (за исключением 1996) имеют среднюю цену значительно ниже 200000.

Заметен значительный разрыв между новыми и старыми машинами: существует резкий переход в цене. Машины, выпущенные после 2009 года, в среднем значительно дороже тех, что были выпущены до 2009 года.

Также видна аномалия 1996 года, он выделяется из общего тренда. Его средняя цена (225000) необычно высока для своего возраста и даже превосходит многие более новые годы. Ожидаемо, самые низкие средние цены наблюдаются у товаров, выпущенных в девяностых и начале двухтысячных годов.


---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 4

*`Сводная таблица (pivot_table) - среднее цена для каждого типа
владельца - столбцы и типа продавца- строки. Отсортировать по возрастанию
seller_type. Округлить до двух знаков.`*

In [33]:
# выполните сводную таблицу согласно варианту
pivot_table = df.pivot_table(index=['seller_type'], columns='owner_type', values='selling_price', aggfunc='mean')

# Сортировка и округление значений сводной таблицы
pivot_table = pivot_table.sort_index(ascending=True)
pivot_table = pivot_table.round(2)
display(pivot_table)

C:\Users\fedor\AppData\Local\Temp\ipykernel_22284\2181757201.py:2: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot_table = df.pivot_table(index=['seller_type'], columns='owner_type', values='selling_price', aggfunc='mean')


owner_type        First Owner  Fourth & Above Owner  Second Owner  \
seller_type                                                         
Dealer           679505.08000          117499.50000  468499.94000   
Individual       518508.08000          182958.88000  326713.81000   
Trustmark Dealer 823709.68000                   NaN  800000.00000   

owner_type        Test Drive Car  Third Owner  
seller_type                                    
Dealer              954293.94000 462333.22000  
Individual                   NaN 259836.07000  
Trustmark Dealer             NaN          NaN

**`Сделайте вывод - интерпретацию - что можно увидеть из результата выполнения данного кода?`**

- Автомобили от первого владельца, имеют самую высокую среднюю стоимость продажи среди всех категорий, за исключением тест-драйв автомобилей. Это логично, так как такие автомобили обычно новее и находятся в лучшем состоянии. 

- Автомобили, у которых было четыре или более владельцев, имеют значительно более низкую среднюю стоимость продажи. Это ожидаемо, так как такие автомобили, вероятно, старше и/или имеют больший износ. Средняя цена для тестовых автомобилей, продаваемых дилерами, очень высока (954293). Это может указывать на то, что это были почти новые модели или модели высокого класса.

- В среднем самые высокие цены у Trustmark дилеров, далее идут обычные дилеры, у которых тоже дорогие машины и затем индивидуальные продавцы, у которых самые низкие цены. 

- Видно, что trustmark дилеры продают только автомобили с одним или двумя владельцами, а тест-драйв машины продают только обычные дилеры.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

### Вывод


***`Сделать расширенный вывод. `***

 В ходе выполнения лабораторной работы был проведен предварительный анализ и предобработка набора данных auto.csv, содержащего информацию о поддержанных автомобилях, выставленных на продажу.

#### На этапе предобработки данных были выполнены следующие шаги:
Стандартизация названий столбцов: все названия были приведены к единому стилю. Обработка пропусков: Пропущенные значения в числовых столбцах были заменены медианными значениями, а в категориальных данных были заменены модой. Удаление дубликатов: были устранены как полные, так и неявные дубликаты. Была проведена оптимизация типов данных
#### Анализ данных с помощью группировок и сводных таблиц позволил сделать следующие выводы:
-	На рынке преобладают автомобили с механической коробкой передач, особенно среди машин с одним или двумя владельцами. Автомобили на газу (LPG/CNG) представлены только с механической коробкой передач, в то время как немногочисленные электромобили — только с автоматической.
-	Существует прямая зависимость между годом выпуска автомобиля и его средней ценой: чем новее автомобиль, тем он дороже. Наблюдается заметный скачок цен для автомобилей, выпущенных после 2009 года (не считая аномалию 1996 года). 
-	Количество предыдущих владельцев и тип продавца значительно влияют на стоимость. Наиболее дорогие автомобили продаются дилерами категории "Trustmark" и имеют не более двух владельцев. Самые низкие цены у частных продавцов.



### Дополнительное задание

**`Подробная формулировка задания`**

Задание 10. Создать столбец “Категория стоимости (price)” (с помощью категоризации).
Выделить минимум 3 категории (низкая, высокая, средняя), фильтрацию для
уровня цены выбрать самостоятельно, аргументировать выбор. Создать
сводную таблицу: средняя и медианная price по категории цены и seller_type.

In [37]:
# код выполнения задания 10
# Объявление границ для разделения на категории по квантилям
low = df['selling_price'].quantile(0.33)
high = df['selling_price'].quantile(0.67)

# Функция для присвоения категории 
def get_category_selling_price(selling_price):
    if selling_price <= low:
        return 'Низкая'
    elif selling_price <= high:
        return 'Средняя'
    else:
        return 'Высокая'

# Создание столбца с типом категории
df['category_price'] = df['selling_price'].apply(get_category_selling_price)
# Сводная таблица
pivot_table =  df.pivot_table(index=['category_price', 'seller_type'], values='selling_price', aggfunc=['mean', 'median'])

display(pivot_table)

C:\Users\fedor\AppData\Local\Temp\ipykernel_22284\3865361383.py:18: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot_table =  df.pivot_table(index=['category_price', 'seller_type'], values='selling_price', aggfunc=['mean', 'median'])
C:\Users\fedor\AppData\Local\Temp\ipykernel_22284\3865361383.py:18: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot_table =  df.pivot_table(index=['category_price', 'seller_type'], values='selling_price', aggfunc=['mean', 'median'])


mean        median
                                selling_price selling_price
category_price seller_type                                 
Высокая        Dealer           1030613.06250  750000.00000
               Individual        893493.49352  708500.00000
               Trustmark Dealer  937115.38462  765000.00000
Низкая         Dealer            187641.61667  200000.00000
               Individual        151929.10799  150000.00000
Средняя        Dealer            374683.52734  375000.00000
               Individual        369528.10173  350000.00000
               Trustmark Dealer  395714.28571  415000.00000

#### Вывод по заданию 10

Для определеняи границ были взять два квантиля, первый в 33 процента, второй в 66 процентов, таким образом получилось разделить значения на оптимальные отрезки. Далее была создана функция дял присвоения категории по цене продажи, если меньше нижнего квантиля, то цена низкая, если больше верхнего квантиля то высокая, если между ними то средняя. И затем был создан столбец категория цены, который высчитывался применяя вышеописанную функцию. И в конце была создана сводная таблица по заданию которая считает среднее и медианное значения в разных группах.

- Заметны выбросы в "Высокой" категории, средняя цена значительно превышает медианную, что указывает на наличие очень дорогих автомобилей (выбросов), искажающих среднее значение.
- В то же время в "Низкой" и "Средней" категориях средняя и медианная цены близки, что говорит о более равномерном распределении цен.
- Профессиональные продавцы (Dealer, Trustmark Dealer) доминируют в сегменте дорогих автомобилей и предлагают более высокие цены в целом.
- Частные продавцы (Individual) предлагают более доступные варианты, особенно в низком и среднем ценовых сегментах.

**`Подробная формулировка задания`**

Задание 1. Создать столбец “Категория пробега (kmdriven)” (с помощью категоризации).
Выделить минимум 3 категории (маленький, большой, средний), фильтрацию
для уровня пробега выбрать самостоятельно, аргументировать выбор. Создать
сводную таблицу: средний и медианный пробег по категории пробега и типу
топлива

In [38]:
# код выполнения задания 1

# Объявление границ для разделения на категории по квантилям
low = df['km_driven'].quantile(0.33)
high = df['km_driven'].quantile(0.67)

# Функция для присвоения категории 
def get_category_mileage(km):
    if km <= low:
        return 'Маленький'
    elif km <= high:
        return 'Средний'
    else:
        return 'Большой'

# Создание столбца с типом категории
df['category_mileage'] = df['km_driven'].apply(get_category_mileage)
# Сводная таблица
pivot_table =  df.pivot_table(index=['category_mileage', 'fuel_type'], values='km_driven', aggfunc=['mean', 'median'])

display(pivot_table)

C:\Users\fedor\AppData\Local\Temp\ipykernel_22284\1035348245.py:19: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot_table =  df.pivot_table(index=['category_mileage', 'fuel_type'], values='km_driven', aggfunc=['mean', 'median'])
C:\Users\fedor\AppData\Local\Temp\ipykernel_22284\1035348245.py:19: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot_table =  df.pivot_table(index=['category_mileage', 'fuel_type'], values='km_driven', aggfunc=['mean', 'median'])


mean       median
                              km_driven    km_driven
category_mileage fuel_type                          
Большой          CNG       100583.44444 100000.00000
                 Diesel    125358.53283 112539.00000
                 LPG       113661.53846 100000.00000
                 Petrol    111704.77814 100000.00000
Маленький        CNG        20012.50000  12050.00000
                 Diesel     27639.35101  30000.00000
                 LPG        32000.00000  32000.00000
                 Petrol     24222.21521  25000.00000
Средний          CNG        69851.95000  70500.00000
                 Diesel     65079.95759  65000.00000
                 Electric   62000.00000  62000.00000
                 LPG        61428.57143  60000.00000
                 Petrol     62806.06329  60000.00000

#### Вывод по заданию 1

Была выбрана фильтрация с использованием квантилей и разделением на 3 равные части, но в этом случае можно использовать и ручное разделение, если есть представление о том, когда у машины уже высокий или еще низкий пробег. Далее аналогично с 10 заданием была создана функция присвоения категории, которая была использована при создании столбца категория пробега. И затем была создана сводная таблица со средними и медианными значениями пробега по группам из задания.

Видна стабильность распределения. Средний и медианный пробег очень близки во всех категориях и для большинства видов топлива, что указывает на равномерное распределение данных без значительных выбросов (за исключением низкой категории с топливом CNG там разница весьма заметна, из-за малого количества записей с этим видом топлива)

- Дизельные автомобили имеют самый высокий пробег в категории "Большой", это подтверждает их использование для длительных поездок, например для фур, которые возят грузы.
- Электромобили представлены только в "Средней" категории, это говорит об их относительной новизне на рынке подержанных авто
- Другие типы топлива показывают схожие пробеги в пределах своих категорий.

**`Подробная формулировка задания`**

Задание 19. Создать столбец “Категория года (year)” (с помощью категоризации).
Выделить минимум 3 категории (старая, новая, средняя), фильтрацию для года
выбрать самостоятельно, аргументировать выбор. Создать столбец “Категория
стоимости (price)” (с помощью категоризации). Выделить минимум 3
категории (низкая, высокая, средняя), фильтрацию для уровня цены выбрать
самостоятельно, аргументировать выбор. Создать сводную таблицу: средний,
медианный пробег по категории года и категории стоимости.


In [40]:
# код выполнения задания 19

# Объявление границ для разделения на категории
# Способ через ручной ввод
low_year = 2009
high_year = 2016
# Способ через квантили
low_selling = df['selling_price'].quantile(0.33)
high_selling = df['selling_price'].quantile(0.67)

# Функция для присвоения категории  цены
def get_category_selling(selling_price):
    if selling_price <= low_selling:
        return 'Низкая'
    elif selling_price <= high_selling:
        return 'Средняя'
    else:
        return 'Высокая'
    
# Функция для присвоения категории года
def get_category_year(year):
    if year <= low_year:
        return 'Старая'
    elif year <= high_year:
        return 'Средняя'
    else:
        return 'Новая'

# Создание столбцов с типом категории
df['category_price'] = df['selling_price'].apply(get_category_selling)
df['category_year'] = df['release_year'].apply(get_category_year)

# Сводная таблица
pivot_table =  df.pivot_table(index=['category_price', 'category_year'], values='km_driven', aggfunc=['mean', 'median'])

display(pivot_table)

mean       median
                                km_driven    km_driven
category_price category_year                          
Высокая        Новая          32048.08870  25000.00000
               Средняя        77406.81667  70000.00000
               Старая         91041.10526  80000.00000
Низкая         Новая          34220.88889  30000.00000
               Средняя        77114.06571  70000.00000
               Старая         89613.82471  80000.00000
Средняя        Новая          30986.31092  25000.00000
               Средняя        76735.89850  70000.00000
               Старая        109237.96341 100000.00000

#### Выводы по заданию 19

Разделение по году было выбрано из-за анализа в прошлых заданиях, видно что резкие изменений в цене начинаются в 2009 и 2016 годах, поэтому было решено взять именно эти года, а для цены использовались квантили, которые разделяют все на равные части. Далее аналогично с прошлыми заданиями созданы функции с присвоением категорий, которые были применены для создания столбцов с этими категориями. А в конце была создана сводная таблица в которой используются обе созданные категории и сравниваются по среднему и медианному пробегу

Наблюдается очевидная прямая зависимость возраста и пробега: чем старше автомобиль, тем выше его пробег

- Для "Новых" и "Средних" по году автомобилей категория стоимости почти не влияет на пробег — он остается соответственно низким или средним.
- Наибольший пробег (109 тыс. км) наблюдается у "Старых" автомобилей в "Средней" ценовой категории. Это указывает на то, что такие автомобили, активно эксплуатировались и представляют собой наиболее массовый сегмент подержанных авто.
- Видна близость среднего и медианного пробега, В большинстве категорий среднее и медианное значения пробега близки, что говорит об отсутствии сильных выбросов. Небольшое расхождение в категории "Старая" и "Средняя" цена может указывать на наличие нескольких автомобилей с очень большим пробегом, которые немного тянут среднее вверх.

**`Подробная формулировка задания`**

Задание 25. Создать столбец “Категория стоимости (price)” (с помощью категоризации).
Выделить минимум 3 категории (низкая, высокая, средняя), фильтрацию для
уровня цены выбрать самостоятельно, аргументировать выбор. Отфильтровать
набор данных. Выбрать только записи, где категория цены= средняя+ выбрать
2 типа топлива с минимальной медианной стоимостью + transmission c
максимальным количеством записей + owner с максимальным количеством
записей. Создать группировки : средний, максимальный, минимальный,
медианный пробег и средняя, максимальная, минимальная, медианная цена по
году и по типу seller_type.



In [41]:
# код выполнения задания 25

# Объявление границ для разделения на категории по квантилям
low_selling_price = df['selling_price'].quantile(0.33)
high_selling_price = df['selling_price'].quantile(0.67)

# Функция для присовения категории
def get_category_selling(selling_price):
    if selling_price <= low_selling_price:
        return 'Низкая'
    elif selling_price <= high_selling_price:
        return 'Средняя'
    else:
        return 'Высокая'

# Создание столбца с типом цены
df['category_price'] = df['selling_price'].apply(get_category_selling)

# Фильтрация
filtered_df = df[df['category_price'] == 'Средняя'].copy()
median_price_by_fuel = filtered_df.groupby('fuel_type')['selling_price'].median().sort_values()
top_2_fuel_types = median_price_by_fuel.iloc[:2].index.tolist()
most_transmission = filtered_df['transmission'].value_counts().idxmax()
most_owner = filtered_df['owner_type'].value_counts().idxmax()

# Применение всех фильтров к датафрейму
df_final_filtered = df[
    (df['category_price'] == 'Средняя') &
    (df['fuel_type'].isin(top_2_fuel_types)) &
    (df['transmission'] == most_transmission) &
    (df['owner_type'] == most_owner)].copy()

# Создание группировки данных
grouped_data = df_final_filtered.groupby(['release_year', 'seller_type']).agg(
    mean_km_driven=('km_driven', 'mean'),
    max_km_driven=('km_driven', 'max'),
    min_km_driven=('km_driven', 'min'),
    median_km_driven=('km_driven', 'median'),
    mean_selling_price=('selling_price', 'mean'),
    max_selling_price=('selling_price', 'max'),
    min_selling_price=('selling_price', 'min'),
    median_selling_price=('selling_price', 'median')
)

display(grouped_data)

C:\Users\fedor\AppData\Local\Temp\ipykernel_22284\2674500925.py:21: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  median_price_by_fuel = filtered_df.groupby('fuel_type')['selling_price'].median().sort_values()
C:\Users\fedor\AppData\Local\Temp\ipykernel_22284\2674500925.py:34: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  grouped_data = df_final_filtered.groupby(['release_year', 'seller_type']).agg(


mean_km_driven  max_km_driven  min_km_driven  \
release_year seller_type                                                      
2014         Dealer                       NaN            NaN            NaN   
             Individual           29000.00000    29000.00000    29000.00000   
             Trustmark Dealer             NaN            NaN            NaN   

                               median_km_driven  mean_selling_price  \
release_year seller_type                                              
2014         Dealer                         NaN                 NaN   
             Individual             29000.00000        290000.00000   
             Trustmark Dealer               NaN                 NaN   

                               max_selling_price  min_selling_price  \
release_year seller_type                                              
2014         Dealer                          NaN                NaN   
             Individual             290000.00000       290000.00000   
             Trustmark Dealer                NaN                NaN   

                               median_selling_price  
release_year seller_type                             
2014         Dealer                             NaN  
             Individual                290000.00000  
             Trustmark Dealer                   NaN

#### Выводы по заданию 25
Для разделения цены продажи на категории использовались квантили. Далее аналогично с прошлыми заданиями создана функция с присвоением категорий, которая была применена для создания столбца с этой категорией. Далее была выполнена фильтрация данных, которая включает в себя выбор только тех записей, где категория цены равна средней, выбраны 2 типа топлива с минимальной медианной стоимостью, тип коробки передач с максимальным количеством записей и тип владельца с макисмальным количеством записей. Затем были созданы группировки: средний, максимальный, минимальный, медианный пробег и средняя, максимальная, минимальная, медианная цена по
году и по типу продавца.

После всех условий фильтрации осталось лишь одно значение с пробегом в 29000 км и ценой в 290000. По данным одной машины не получится провести анализ с полезной и значимой информацией. Это показывает, что условия отбора оказались слишком строгими